modin/experimental/batch/test/test_pipeline.py

```diff
@@ -0,0 +1,733 @@
+    [ptn.drain_call_queue() for ptn in ptns]
+    ptns = [ptn.list_of_blocks for ptn in ptns]
+    corr_df = from_partitions(ptns, axis=None)
+    # Compare pipelined and non-pipelined df
+    df_equals(corr_df, new_df)
+    # Ensure that setting `num_partitions` when creating a pipeline changes `NPartitions`
+    num_ptns = NPartitions.get()
+    df = df._build_batch_pipeline(lambda df: df, 0, num_partitions=(num_ptns - 1))
+    assert NPartitions.get() == (
+        num_ptns - 1
+    ), "Pipeline did not change NPartitions.get()"
+    NPartitions.put(num_ptns)
+
+
+@pytest.mark.skipif(
+    Engine.get() == "Ray",
+    reason="Ray supports the Batch Pipeline API",
+)
+def test_pipeline_unsupported_engine():
+    """
+    This test ensures that trying to use the Pipeline API with an unsupported Engine raises errors.
+    """
+    from modin.experimental.batch.pipeline import PandasQueryPipeline
+
+    # Check that top level API does not allow `Engine` to not be Ray.
+    df = pd.DataFrame([[1, 2, 3], [4, 5, 6]])
+    with pytest.raises(
+        NotImplementedError,
+        match="Batch Pipeline API is only implemented for Ray Engine.",
+    ):
+        df = df._build_batch_pipeline(lambda df: df, 0)
+
+    # Check that PandasQueryPipeline object checks that Engine is Ray.
+    with pytest.raises(
+        NotImplementedError,
+        match="Batch Pipeline API is only implemented for Ray Engine.",
+    ):
+        PandasQueryPipeline(df, 0)
+    eng = Engine.get()
+    Engine.put("Ray")
+    # Check that even if Engine is Ray, if the df is not backed by Ray, the Pipeline does not allow initialization.
+    with pytest.raises(
+        NotImplementedError,
+        match="Batch Pipeline API is only implemented for Ray Engine.",
+    ):
+        df = df._build_batch_pipeline(lambda df: df, 0)
+    Engine.put(eng)
+
+
+@pytest.mark.skipif(
+    Engine.get() != "Ray",
+    reason="Only Ray supports the Batch Pipeline API",
+)
+def test_pipeline_multiple_outputs():
+    """
+    This test creates a pipeline with multiple outputs, and checks that all are computed correctly.
+    """
+    arr = np.random.randint(0, 1000, (1000, 1000))
+    df = pd.DataFrame(arr)
+    df = df._build_batch_pipeline(lambda df: df * -30, 0, is_output=True)
+    df = df._add_batch_query(
+        lambda df: df.rename(columns={i: f"col {i}" for i in range(1000)}),
+        is_output=True,
+    )
+    df = df._add_batch_query(lambda df: df + 30, is_output=True)
+    new_dfs = df._compute_batch()
+    assert len(new_dfs) == 3, "Pipeline did not return all outputs"
+    corr_df = pd.DataFrame(arr) * -30
+    df_equals(corr_df, new_dfs[0])  # First output computed correctly
+    corr_df = corr_df.rename(columns={i: f"col {i}" for i in range(1000)})
+    df_equals(corr_df, new_dfs[1])  # Second output computed correctly
+    corr_df += 30
+    df_equals(corr_df, new_dfs[2])  # Third output computed correctly
+    # Testing Output ID + Multiple Outputs now
```

Comment:
I think we should make this a separate test case.

Reply:
To confirm - you think testing output id with multiple outputs should be a different case?

Reply:
Right.